libraries/standard/mqtt/src/mqtt.c

```diff
@@ -951,10 +965,30 @@ static MQTTStatus_t receiveConnack( const MQTTContext_t * pContext,
                                                       pContext->transportInterface.networkContext,
                                                       pIncomingPacket );
 
-        timeExpired = ( calculateElapsedTime( getTimeStamp(), entryTimeMs ) >= timeoutMs ) ? true : false;
+        /* The loop times out based on 2 conditions.
+         * 1. If timeoutMs is greater than 0:
+         *    Loop times out based on the timeout calculated by getTime()
+         *    function.
+         * 2. If timeoutMs is 0:
+         *    Loop times out based on the maximum number of retries config
+         *    MQTT_MAX_CONNACK_RECEIVE_RETRY_COUNT. This config will control
+         *    maximum the number of retry attempts to read the CONNACK packet.
+         *    A value of 0 for the config will try once to read CONNACK. */
+        useRetry = ( timeoutMs == 0U ) ? true : false;
+
+        if( useRetry == false )
+        {
+            /* Loop runs only once for 0 timeoutMs. */
+            breakFromLoop = ( calculateElapsedTime( getTimeStamp(), entryTimeMs ) >= timeoutMs ) ? true : false;
+        }
+        else
+        {
+            breakFromLoop = ( loopCount >= MQTT_MAX_CONNACK_RECEIVE_RETRY_COUNT ) ? true : false;
+            loopCount++;
+        }
```

Comment:
Could the `useRetry` flag be avoided with the following?
```suggestion
        /* If timeout is passed, use timeout value to determine whether we should retry reception of CONNACK. */
        if( timeout == 1u )
        {
            /* Loop runs only once for 0 timeoutMs. */
            breakFromLoop = ( calculateElapsedTime( getTimeStamp(), entryTimeMs ) >= timeoutMs ) ? true : false;
        }
        /* Use the configured retry count to determine if we should retry reception of CONNACK. */
        else
        {
            breakFromLoop = ( loopCount >= MQTT_MAX_CONNACK_RECEIVE_RETRY_COUNT ) ? true : false;
            loopCount++;
        }
```

Reply:
Yea. Thanks. Added it for an additional logic and later removed. Will make that fix.